src_c/draw.c

```diff
@@ -1564,114 +1694,216 @@ draw_arc(SDL_Surface *dst, int x, int y, int radius1, int radius2,
+ * with additional line width parameter
  */
 static void
-draw_circle_pixel(SDL_Surface *dst, int x0, int y0, int x1, int y1, Uint32 color)
+draw_circle_bresenham(SDL_Surface *dst, int x0, int y0, int radius, int thickness, Uint32 color, int *drawn_area)
 {
-    // (x0, y0) -> center of the circle, (x1, y1) -> pixel to bve drawn
-    // leading_bit -> used for fast quadrant calculation, 1 for negative and 0 for positive
-    // number (number is "vector" from center to the pixel)
-    int x_leading_bit = ((x1-x0) > 0) - ((x1-x0) < 0) == 1 ? 0 : 1;
-    int y_leading_bit = ((y1-y0) > 0) - ((y1-y0) < 0) == 1 ? 0 : 1;
-    int quadrant = (x_leading_bit != y_leading_bit) + y_leading_bit + y_leading_bit + 1;
-    if (quadrant == 1) {
-        set_at(dst, x1 - 1, y1 - 1, color);  // Move one to the left and up
-    }
-    else if (quadrant == 2) {
-        set_at(dst, x1, y1 - 1, color);      // Move one to the up
-    }
-    else if (quadrant == 3) {
-        set_at(dst, x1, y1, color);
-    }
-    else {
-        set_at(dst, x1 - 1, y1, color);      // Move one to the left
+    int f = 1 - radius;
+    int ddF_x = 0;
+    int ddF_y = -2 * radius;
+    int x = 0;
+    int y = radius;
+    int y1;
+    int i_y = radius-thickness;
+    int i_f = 1 - i_y;
+    int i_ddF_x = 0;
+    int i_ddF_y = -2 * i_y;
+    int i;
+
+    while(x < y) {
+        if(f >= 0) {
+            y--;
+            ddF_y += 2;
+            f += ddF_y;
+        }
+        if(i_f >= 0)
+        {
+            i_y--;
+            i_ddF_y += 2;
+            i_f += i_ddF_y;
+        }
+        x++;
+        ddF_x += 2;
+        f += ddF_x + 1;
+
+        i_ddF_x += 2;
+        i_f += i_ddF_x + 1;
+
+        if(thickness>1)
+            thickness=y-i_y;
+
+        /* Numbers represent parts of circle function draw in radians
+           interval: [number - 1 * pi / 4, number * pi / 4] */
+        for (i=0; i<thickness; i++){
+            y1=y-i;
+            if ((y0 + y1 - 1) >= (y0 + x - 1))
+                set_at(dst, x0 + x - 1, y0 + y1 - 1, color, drawn_area);  /* 7 */
+            if ((y0 + y1 - 1) >= (y0 + x - 1))
+                set_at(dst, x0 - x, y0 + y1 - 1, color, drawn_area);      /* 6 */
+            if ((y0 - y1) <= (y0 - x))
+                set_at(dst, x0 + x - 1, y0 - y1, color, drawn_area);      /* 2 */
+            if ((y0 - y1) <= (y0 - x))
+                set_at(dst, x0 - x, y0 - y1, color, drawn_area);          /* 3 */
+            if ((x0 + y1 - 1) >= (x0 + x - 1))
+                set_at(dst, x0 + y1 - 1, y0 + x - 1, color, drawn_area);  /* 8 */
+            if ((x0 - y1) <= (x0 - x))
+                set_at(dst, x0 - y1, y0 + x - 1, color, drawn_area);      /* 5 */
+            if ((x0 + y1 - 1) >= (x0 + x - 1))
+                set_at(dst, x0 + y1 - 1, y0 - x, color, drawn_area);      /* 1 */
+            if ((x0 - y1) <= (x0 - x))
+                set_at(dst, x0 - y1, y0 - x, color, drawn_area);          /* 4 */
```

Comment:
Some of these `if` checks are the same. Can they be combined?

Reply:
Good find. JoKing ran out of time. I'm starting another branch to do this.